src/client/dfs/dfs.c

```diff
@@ -319,15 +336,57 @@ fetch_entry(daos_handle_t oh, daos_handle_t th, const char *name, size_t len,
 	if (strcmp(name, ".") == 0)
 		D_ASSERT(0);
 
+	if (xnr) {
+		D_ALLOC_ARRAY(pxnames, xnr);
```

Comment:
@ashleypittman  we can optimize this here by reducing the number of allocations and making this only specific to dfuse as you were suggesting earlier I think, with a predefined array of xattr names (with the x: prefix).
we would need to know the list of all xattrs names being fetched on every lookup in addition to an upper bound on the size of their values.

Reply:
for dfuse the list would be constant for all lookups so we could do that, knowing the max size for user.daos and user.dfuse.uns would be possible, but security.ima is probably not a bounded size however.  99% of time that wouldn't be set, so potentially we could simply check if that one existed although it would complicate the interface.

Another thought I had is that xattrs are only allowed with "system", "user", "security" and "trusted" namespace, so we could check the namespace and reject or only use the x: suffix if it's not one of these four.  We could then avoid the concat and extra memory allocation at all.

https://elixir.bootlin.com/linux/latest/source/fs/xattr.c#L56

Reply:
in the DFS case, i would rather keep all xattr with the x: suffix since this is the format. i don't want to change the format just for what dfuse / POSIX expects.
my suggestion was only for this function for which we know dfuse is the only one using it.
in any case, this optimization can be done in a further change to this API, but this should allow now to revert the mknod and fifo change to DFS.

Reply:
The x: change would affect the durable format only, it wouldn't make any change to the dfs api, so dfuse would not be aware of it.  It was just an idea to avoid the extra malloc call.

Reply:
exactly. a format change i would like to avoid.